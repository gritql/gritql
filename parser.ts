import { argumentsToObject } from './arguments'
import { $typeSymbol, parseDirective } from './directives'
import { transformFilters } from './filters'
import { checkPropTypes, PropTypes } from './types'

export function parseMetric(tree, query, knex) {
  if (query.metricResolvers[tree.name?.value])
    return query.metricResolvers[tree.name?.value](tree, query, knex)
  else return query.metricResolvers.default(tree, query, knex)
}

export function parseDimension(tree, query, knex) {
  if (query.dimensionResolvers[tree.name?.value])
    return query.dimensionResolvers[tree.name?.value](tree, query, knex)
  else return query.dimensionResolvers.default(tree, query, knex)
}

export function parseFilters(tree, query, knex) {
  const { arguments: args } = tree
  return transformFilters(
    args.concat({ name: { value: 'from' }, value: { value: query.table } }),
    query,
    knex,
  )
}

export const defaultTypes = {
  // Basic JS types infered from PropTypes
  // That is enough to build any kind of types
  Any: PropTypes.any,
  Number: PropTypes.number,
  String: PropTypes.string,
  Bool: PropTypes.bool,
  Object: PropTypes.object,
  Array: PropTypes.array,
}

// TODO: No nested array support right now
export function parseType(type, context) {
  if (type.kind === 'UnionTypeDefinition') {
    context.types[type.name.value] = PropTypes.oneOfType(
      type.types.map((t) => {
        if (context.types[t.name.value]) {
          return context.types[t.name.value]
        } else {
          throw new Error(`${t.name.value} is not declared type`)
        }
      }),
    )
  } else if (type.kind === 'TupleTypeDefinition') {
    context.types[type.name.value] = PropTypes.tuple(
      ...type.elements.map((element) => {
        const isRequired = element.kind === 'NonNullType'
        element = element.type || element

        if (context.types[element.name.value]) {
          let type = context.types[element.name.value]

          if (isRequired) {
            type = type.isRequired
          }

          return type
        } else {
          throw new Error(`${element.name.value} is not declared type`)
        }
      }),
    )
  } else if (type.kind === 'EnumTypeDefinition') {
    context.types[type.name.value] = PropTypes.oneOf(
      type.values.map((t) => t.name.value),
    )
  } else if (
    type.kind === 'InputObjectTypeDefinition' ||
    type.kind === 'ObjectTypeDefinition'
  ) {
    const isInput = type.kind.startsWith('Input')

    const shape = type.fields
      .filter(({ kind }) => kind !== $typeSymbol)
      .reduce((acc, field) => {
        const isRequired = field.type.kind === 'NonNullType'

        if (isRequired) {
          field.type = field.type.type
        }

        if (field.type.kind === 'NamedType') {
          if (!context.types[field.type.name.value]) {
            throw new Error(`${field.type.name.value} is not declared type`)
          }

          // TODO: Default value for object type/input field is not implemented yet
          acc[field.name.value] = context.types[field.type.name.value]
        } else if (field.type.kind === 'ListType') {
          field.type = field.type.type

          const isRequired = field.type.kind === 'NonNullType'

          if (isRequired) {
            field.type = field.type.type
          }

          if (!context.types[field.type.name.value]) {
            throw new Error(`${field.type.name.value} is not declared type`)
          }

          acc[field.name.value] = context.types[field.type.name.value]

          if (isRequired) {
            acc[field.name.value] = acc[field.name.value].isRequired
          }

          acc[field.name.value] = PropTypes.arrayOf(acc[field.name.value])
        }

        if (isRequired) {
          acc[field.name.value] = acc[field.name.value].isRequired
        }

        return acc
      }, {})

    const symbolField = type.fields.filter(
      ({ kind }) => kind === $typeSymbol,
    )[0]

    if (type.options?.isMapped && !context.types[symbolField.type.name.value]) {
      throw new Error(`${symbolField.type.name.value} is not declared type`)
    }

    if (type.name.value === 'BrandFilter') {
      console.log(shape, JSON.stringify(type))
    }

    context.types[type.name.value] = type.options?.isMapped
      ? PropTypes.map(
          shape,
          PropTypes[symbolField.key.name.value.toLowerCase()].isRequired,
          context.types[symbolField.type.name.value],
        )
      : PropTypes[isInput ? 'shape' : 'exact'](shape)
  }

  return context
}

export function parseVariableDefinition(def, context) {
  const name = def?.variable?.name?.value
  const isRequired = def.type.kind === 'NonNullType'

  if (isRequired) {
    def.type = def.type.type
  }

  if (def.type.kind === 'NamedType') {
    if (context.types[def.type.name.value]) {
      context.variablesValidator[name] = context.types[def.type.name.value]
    } else {
      throw new Error(`${def.type.name.value} is not declared type`)
    }
  } else if (def.type.kind === 'ListType') {
    def.type = def.type.type

    const isRequired = def.type.kind === 'NonNullType'

    if (isRequired) {
      def.type = def.type.type
    }

    if (context.types[def.type.name.value]) {
      context.variablesValidator[name] = context.types[def.type.name.value]
    } else {
      throw new Error(`${def.type.name.value} is not declared type`)
    }

    if (isRequired) {
      context.variablesValidator[name] =
        context.variablesValidator[name].isRequired
    }

    context.variablesValidator[name] = PropTypes.arrayOf(
      context.variablesValidator[name],
    )

    if (!!def.defaultValue) {
      context.variables[name] =
        context.variables[name] ?? argumentsToObject(def.defaultValue)
    }
  }

  if (isRequired) {
    context.variablesValidator[name] =
      context.variablesValidator[name].isRequired
  }

  return context
}

export function processArguments(args, context) {
  return args
    .map((argument) => {
      if (argument.value?.kind === 'Variable') {
        if (context.variables[argument.value.name.value] !== undefined) {
          return {
            ...argument,
            value: {
              kind: 'JSONValue',
              value: context.variables[argument.value.name.value],
            },
          }
        } else {
          return null
        }
      } else if (argument.value?.kind === 'ObjectValue') {
        argument.value.fields = processArguments(argument.value.fields, context)
      } else if (argument.value?.kind === 'ListValue') {
        argument.value.values = processArguments(argument.value.values, context)
      }

      return argument
    })
    .filter(Boolean)
}

export function processSelections(selections, field, query, context) {
  if (field?.kind === 'FragmentSpread') {
    if (context.fragments[field.name.value]) {
      selections = selections.concat(
        compileFragment(context.fragments[field.name.value], query, []),
      )

      return selections
    } else {
      throw new Error(`${field.name.value} fragment is not defined yet`)
    }
  }

  if (field?.name.value === 'use') {
    const { fragment, args } = {
      args: field.arguments.filter((f) => f.name.value !== 'fragment'),
      fragment: field.arguments.filter((f) => f.name.value === 'fragment'),
    }

    if (fragment.length) {
      if (context.fragments[fragment[0].value.value]) {
        selections = selections.concat(
          compileFragment(
            context.fragments[fragment[0].value.value],
            query,
            args,
          ),
        )

        return selections
      } else {
        throw new Error(`${field.name.value} fragment is not defined yet`)
      }
    } else {
      throw new Error(
        `use requires fragment arguments with name of fragment which you want to compile`,
      )
    }
  }

  if (field?.directives) {
    field.directives = field.directives.map((directive) => {
      if (directive.arguments) {
        directive.arguments = processArguments(directive.arguments, context)
      }

      return directive
    })

    field = parseDirective(field, null, 'field')
  }

  if (field?.arguments) {
    field.arguments = processArguments(field.arguments, context)
  }

  if (field?.selectionSet) {
    field.selectionSet.selections = field.selectionSet.selections
      .reduce(
        (selections, field) =>
          processSelections(selections, field, query, context),
        [],
      )
      .filter(Boolean)
  }

  if (field) {
    selections.push(field)
  }

  return selections
}

export function compileFragment(fragment, query, arglist: any) {
  const args = argumentsToObject(arglist)

  const context = {
    types: query.context.types,
    variables: {
      ...query.context.variables,
      ...args,
    },
    variablesValidator: {},
  }

  if (fragment.variableDefinitions) {
    fragment.variableDefinitions.map((def) => {
      parseVariableDefinition(def, context)
    })

    checkPropTypes(
      context.variablesValidator,
      context.variables,
      'fragment',
      fragment.name.value,
    )
  }

  return fragment.selections
    .reduce(
      (selections, field) =>
        processSelections(selections, field, query, context),
      [],
    )
    .filter(Boolean)
}
