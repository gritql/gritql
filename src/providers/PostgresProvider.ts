import {
  basicSqlDimensions,
  basicSqlMetrics,
  Instruction,
} from '../Instructions/basic'
import { KnexBasedSQLProvider } from './KnexBasedSQLProvider'

export default class PostgresProvider extends KnexBasedSQLProvider {
  name = 'postgres'

  constructor(configuration, connector?) {
    super(configuration, connector || require('postgres'))
  }
  initiateQuery({ table }) {
    const builder = this.getQueryBuilder()
    return { promise: builder.select().from(table), builder }
  }
  instructions = Instructions
  metrics = basicSqlMetrics
  dimensions = basicSqlDimensions

  getMetric(name) {
    const metric = this.metrics.find((instruction) => instruction.name === name)
    return metric
  }
  getDimension(name) {
    const dimension = this.dimensions.find(
      (instruction) => instruction.name === name,
    )
    if (!dimension) return this.dimensions[0] //return default
    return dimension
  }
  getInstruction(name: string) {
    const instruction = this.instructions.find(
      (instruction) => instruction.name === name,
    )
    if (!instruction) {
      throw new Error(`Instruction ${name} not found in ${this.name} provider`)
    }
    return instruction
  }
}

const apply_filters: Instruction = function apply_filters(
  this,
  { promise, builder },
  { args: filters, advancedFilters },
) {
  // todo: refactor, remove advanced filters. All should be normalised if possible.
  if (advancedFilters) {
    if (advancedFilters?.where) {
      promise = promise.where(builder.raw(advancedFilters.where))
    }

    if (advancedFilters?.having) {
      promise = promise.having(builder.raw(advancedFilters.having))
    }
  }
  return filters.reduce((queryPromise, filter, i) => {
    const selector =
      filter[1] === 'in' ? 'whereIn' : i === 0 ? 'where' : 'andWhere'
    return queryPromise[selector].apply(
      queryPromise,
      filter[1] === 'in'
        ? filter.filter((a) => a !== 'in')
        : filter[1] === 'search'
        ? [
            builder.raw(
              `to_tsvector('simple', ??) @@ (plainto_tsquery('simple', ?)::text || ':*')::tsquery`,
              [filter[0], filter[2]],
            ),
          ]
        : filter,
    )
  }, promise)
}

apply_filters.keywords = ['WHERE', 'IN', 'AND', 'OR', 'LIKE', 'ILIKE']
apply_filters.priority = 0

const Instructions: Instruction[] = [apply_filters].concat(basicSqlMetrics)
