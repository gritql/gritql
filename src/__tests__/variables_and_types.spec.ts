import gql from 'graphql-tag'
import { defaultTypes } from '../../parser'
import GritQL from '../gritql'
import PostgresProvider from '../providers/PostgresProvider'

const postgresqlProvider = new PostgresProvider({})
const gritql = new GritQL()
gritql.use(postgresqlProvider)

function getSql(query, variables: any) {
  const defs = gql(query).definitions
  return gritql
    .queryParser(defs, undefined, 0, {
      fragments: {},
      types: { ...defaultTypes },
      variablesValidator: {},
      variables,
      typeDefinitions: {},
    })[0]
    .query.renderQuery()
    .then((context) => {
      return context.promise.toSQL().sql
    })
}

describe('gritql variables && types', () => {
  test('input type must be traited as exact object type', () => {
    getSql(
      `
    input Date {
      gt: String!
      lt: String!
    }

    input Filters {
      brand: String!
      country: String!
      date: Date
    }

    query table($filters: Filters) {
      fetch(filters: $filters) {
        value: sum(a:pageviews)
      }
    }
  `,
      {
        filters: {
          brand: 'Adidas',
          date: {
            gt: '2021-01-01',
            lt: '2022-01-01',
          },
          country: 'US',
        },
      },
    ).then(matchSnapshot)
  })

  test('list type must be traited as array type', () => {
    getSql(
      `
    input Date {
      gt: String!
      lt: String!
    }

    input Filters {
      brand: String!
      country: String!
      date: Date
      categories: [Number!]!
    }

    query table($filters: Filters, $brand: [String!]!) {
      fetch(filters: $filters) {
        value: sum(a:pageviews)
      }
    }
  `,
      {
        filters: {
          brand: 'Adidas',
          date: {
            gt: '2021-01-01',
            lt: '2022-01-01',
          },
          country: 'US',
          categories: [1, 11, 6],
        },
        brand: ['Adidas', 'Nike'],
      },
    ).then(matchSnapshot)
  })

  test('type must be traited as shape object type', () => {
    getSql(
      `
    type Date {
      gt: String!
      lt: String!
    }

    type Filters {
      brand: String!
      country: String!
      date: Date
    }

    query table($filters: Filters) {
      fetch(filters: $filters) {
        value: sum(a:pageviews)
      }
    }
  `,
      {
        filters: {
          brand: 'Adidas',
          date: {
            gt: '2021-01-01',
            lt: '2022-01-01',
          },
          country: 'US',
        },
      },
    ).then(matchSnapshot)
  })

  test('Union type', () => {
    getSql(
      `
    type Date1 {
      gt: String!
      lt: String!
    }

    type Date2 {
      gte: String!
      lte: String!
    }

    union Date = Date1 | Date2

    type Filters {
      brand: String!
      country: String!
      date: Date
    }

    query table($filters: Filters) {
      fetch(filters: $filters) {
        value: sum(a:pageviews)
      }
    }
  `,
      {
        filters: {
          brand: 'Adidas',
          date: {
            gt: '2021-01-01',
            lt: '2022-01-01',
          },
          country: 'US',
        },
      },
    ).then(matchSnapshot)
  })
  //TODO: fix this one, seems doesn't work properly
  test('Custom types directives', () => {
    getSql(
      `
    type YearDate {
      year: Number!
    }

    type MonthDate @inherits(name: YearDate) {
      month: Number!
    }

    type DayDate @inherits(name: MonthDate) {
      day: Number!
    }

    type TupleDate @tuple(definitions: [DayDate, DayDate], isRequired: true)

    enum Operators {
      eq
      ilike
      like
    }

    enum NumOperators {
      lt
      gt
      gte
      lte
    }

    type BrandOps @map(key: Operators, value: String) @map(key: NumOperators, value: Number)

    type BrandFilter @inherits(name: BrandOps) @map(key: String, value: BrandOps)

    input DateEqFilter {
      eq: TupleDate
    }

    type Filters {
      or: [BrandFilter!]!
      country: String!
      date: DateEqFilter
    }

    query table($filters: Filters) {
      fetch(filters: $filters) {
        value: sum(a:pageviews)
      }
    }
  `,
      {
        filters: {
          or: [{ brand: { eq: 'Adidas' }, price: { lt: 2 } }],
          date: {
            eq: [
              { year: 2021, month: 12, day: 31 },
              { year: 2022, month: 9, day: 1 },
            ],
          },
          country: 'US',
        },
      },
    ).then(matchSnapshot)
  })
})

function matchSnapshot(sql) {
  expect(sql).toMatchSnapshot()
}
