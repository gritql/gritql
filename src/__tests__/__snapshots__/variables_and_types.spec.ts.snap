// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gritql variables && types Custom types directives 1`] = `"select sum("pageviews") from "table" where ((((("table"."brand" = 'Adidas')) AND (("table"."price" < 2))))) AND (("table"."date" = E'{"{\\\\"year\\\\":2021,\\\\"month\\\\":12,\\\\"day\\\\":31}","{\\\\"year\\\\":2022,\\\\"month\\\\":9,\\\\"day\\\\":1}"}')) AND ("table"."country" = 'US')"`;

exports[`gritql variables && types Union type 1`] = `"select sum("pageviews") from "table" where ("table"."brand" = 'Adidas') AND (("table"."date" > '2021-01-01') AND ("table"."date" < '2022-01-01')) AND ("table"."country" = 'US')"`;

exports[`gritql variables && types input type must be traited as exact object type 1`] = `"select sum("pageviews") from "table" where ("table"."brand" = 'Adidas') AND (("table"."date" > '2021-01-01') AND ("table"."date" < '2022-01-01')) AND ("table"."country" = 'US')"`;

exports[`gritql variables && types list type must be traited as array type 1`] = `"select sum("pageviews") from "table" where ("table"."brand" = 'Adidas') AND (("table"."date" > '2021-01-01') AND ("table"."date" < '2022-01-01')) AND ("table"."country" = 'US') AND (("table"."categories" = 1) AND ("table"."categories" = 11) AND ("table"."categories" = 6))"`;

exports[`gritql variables && types type must be traited as shape object type 1`] = `"select sum("pageviews") from "table" where ("table"."brand" = 'Adidas') AND (("table"."date" > '2021-01-01') AND ("table"."date" < '2022-01-01')) AND ("table"."country" = 'US')"`;
